src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs

```diff
@@ -75,7 +74,17 @@ public static partial class JsonSerializer
                 }
                 else if (tokenType == JsonTokenType.StartObject)
                 {
-                    HandleStartObject(options, ref reader, ref state);
+                    if (!state.Current.IsProcessingProperty)
+                    {
+                        HandleStartObject(options, ref reader, ref state);
+                    }
+                    else
+                    {
+                        if (HandleValue(tokenType, options, ref reader, ref state))
+                        {
+                            return;
```

Comment:
Why do we return here but not in the other branch?

Reply:
Because the existing branch goes back to the `reader.Read()` loop unconditionally. It makes for awkward reading, I know. I'm giving thought as to how to make this easier to read- I'll have to muck with this a bit to handle the `Stream` issue.